rfd/0061-tsh-aliases.md

```diff
@@ -0,0 +1,165 @@
+state: draft
+---
+
+# RFD 61 - tsh command aliases
+
+## What
+
+Proposes a way for users to define custom tsh commands.
+
+## Why
+
+Defining custom commands will allow users to extend `tsh` CLI to fit their
+use-case without having to modify Teleport's source code.
+
+Examples of what users might want to do include:
+
+- Define shorthands for the existing `tsh` commands.
+
+- Tweak the default behavior of existing `tsh` commands, for example to have
+  `tsh login` always default to a particular auth connector or username.
+
+- Define new subcommands combining multiple actions in the same command, for
+  example logging into a cluster, a database and connecting to the database.
+
+## Prior art
+
+### bash aliases
+
+The simplest analogue that comes to mind is the bash alias:
+
+```bash
+alias ll="ls -l"
+```
+
+Bash aliases are quite simplistic as they only allow to define new commands
+that expand to the strings they alias and don't provide a way to refer to the
+command's arguments.
+
+### git aliases
+
+Git aliases are defined in the repo-specific `.gitconfig` (or `.git/config`)
+or global `/etc/gitconfig` file, and are much closer conceptually to what we
+want to accomplish as they allow to define custom `git` subcommands:
+
+```ini
+[alias]
+co = checkout
+ls = log --oneline
+```
+
+Git aliases also support calling non-`git` commands (using `!` notation) and
+refer to the command's arguments (via `$@`, `$1`, etc.) which allows to build
+powerful new commands.
+
+For example, the following alias will make `git top` to show 15 most recently
+updated local branches:
+
+```ini
+[alias]
+top = "!bash -c 'git branch --sort=-committerdate | head -15'"
+```
+
+## Configuration and examples
+
+Users will define aliases in their `tsh` configuration file which is kept
+in `$TELEPORT_HOME/config/config.yaml` using the following syntax:
+
+```yaml
+aliases:
+    "<alias>": "<command>"
+```
+
+The `<alias>` can only be a top-level subcommand. In other words, users can
+define a new `tsh mycommand` alias but cannot define `tsh my command` command.
+
+A few notes regarding the `<command>`:
+
+- Unlike `git` aliases explored above, proposal is to always require the binary
+  name instead of defaulting to `tsh` and requiring special syntax otherwise.
+- The `<command>` will be fed to `exec.Command` for execution. To chain multiple
+  commands, use pipes, etc. users will explicitly use `bash -c`.
+- The `<command>` will require `$@`, `$1`, etc. to reference the arguments
+  provided to the alias command.
+
+### Examples
+
+Starting with a very simple example, suppose you're tired of typing `tsh login`
+every morning and want to alias it to something shorter e.g. `tsh l`:
+
+```yaml
+aliases:
+    "l": "tsh login $@"
+```
+
+As a more practical example, the following alias will make `tsh login` default
+to the specific auth connector and username:
+
+```yaml
+aliases:
+    "login": "tsh login --auth=local --user=alice $@"
+```
```

Comment:
Do you think it might be handy to have a way to temporarily disable alias? Otherwise need to remove the alias or use a different teleport home if want to login someone or somewhere else (like one time thing) using this example. Or, git alias for the same command does not take any effect so it always run the one from the binary.

Reply:
Yeah, for one-offs like this I think running with `TELEPORT_HOME=/tmp/tsh tsh xxx` or temporarily disabling the alias is probably the way to go, not sure it deserves another flag tbh.